File: lclang/src/main/java/lclang/statements/WhileStatement.java
package lclang.statements;

import lclang.Caller;
import lclang.LCBaseExecutor;
import lclang.Value;
import lclang.exceptions.LCLangRuntimeException;
import lclang.libs.lang.classes.BoolClass;
import lclang.libs.lang.classes.VoidClass;
import lclang.statements.expressions.Expression;

public class WhileStatement extends Statement {
    public final Expression condition;
    public final Statement body;

    public WhileStatement(Expression condition, Statement body) {
        super(0);
        this.condition = condition;
        this.body = body;
    }

    @Override
    public Value visit(Caller prevCaller, LCBaseExecutor visitor) throws LCLangRuntimeException {
        if(body==null) {
            while(condition.visit(prevCaller, visitor).get.invoke(prevCaller) != BoolClass.FALSE);
        }else while(condition.visit(prevCaller, visitor).get.invoke(prevCaller) != BoolClass.FALSE) {
            Value value = body.visit(prevCaller, visitor);
            if(value.state==Value.State.RETURN)
                return value;
            else if(value.state==Value.State.STOP)
                break;
        }

        return VoidClass.value;
    }
}

File: lclang/src/main/kotlin/lclang/statements/WhileStatement.kt
package lclang.statements

import lclang.Caller
import lclang.LCBaseExecutor
import lclang.VOID_VALUE
import lclang.Value
import lclang.libs.lang.BoolClass

class WhileStatement(
    val condition: Expression,
    val body: Statement?,
    line: Int,
    column: Int
): Statement(line, column) {
    override fun visit(prevCaller: Caller, visitor: LCBaseExecutor): Value {
        val caller = getCaller(prevCaller)
        if(body==null) {
            while(condition.visit(caller, visitor).get(caller)!= BoolClass.FALSE) {}
        }else while(true) {
            if(condition.visit(caller, visitor).get(caller)== BoolClass.FALSE) break

            val value = body.visit(caller, visitor)
            if(value.state==Value.State.RETURN)
                return value
            else if(value.state==Value.State.STOP)
                break
        }

        return VOID_VALUE
    }
}